alephium/lib/deployment.ts

```diff
@@ -0,0 +1,248 @@
+}
+
+export type DeployFunction = (deployer: Deployer, networkType: NetworkType) => Promise<void>
+
+export class PrivateKeySigner extends SignerWithNodeProvider {
+  private static ec = new EC('secp256k1')
+
+  private readonly privateKey: string
+  private readonly publicKey: string
+  private readonly address: string
+  private readonly group: number
+
+  constructor(provider: NodeProvider, privateKey: string, alwaysSubmitTx = true) {
+    super(provider, alwaysSubmitTx)
+    this.privateKey = privateKey
+    this.publicKey = publicKeyFromPrivateKey(privateKey)
+    this.address = addressFromPublicKey(this.publicKey)
+    this.group = groupOfAddress(this.address)
+  }
+
+  static from(mnemonic: string, providerUrl: string): PrivateKeySigner {
+    const wallet = getWalletFromMnemonic(mnemonic)
+    const provider = new NodeProvider(providerUrl)
+    return new PrivateKeySigner(provider, wallet.privateKey)
+  }
+
+  getAccountSync(): Account {
+    return { address: this.address, publicKey: this.publicKey, group: this.group }
+  }
+
+  async getAccounts(): Promise<Account[]> {
+    return [this.getAccountSync()]
+  }
+
+  async signRaw(signerAddress: string, hexString: string): Promise<string> {
+    if (signerAddress !== this.address) {
+      throw Error('Unmatched signer address')
+    }
+
+    const key = PrivateKeySigner.ec.keyFromPrivate(this.privateKey)
+    const signature = key.sign(hexString)
+    return signatureEncode(signature)
+  }
+}
+
+function createDeployer(network: Network, environments: Map<string, string>): Deployer {
+  const signer = PrivateKeySigner.from(network.mnemonic, network.nodeUrl)
+  const account = signer.getAccountSync()
+
+  const deployContract = async (contract: Contract, params: DeployContractParams): Promise<DeployContractResult> => {
+    const result = await contract.transactionForDeployment(signer, params)
+    await signer.submitTransaction(result.unsignedTx, result.txId)
+    const confirmed = await waitTxConfirmed(signer.provider, result.txId)
+    return {...result, blockHash: confirmed.blockHash}
+  }
+
+  const runScript = async (script: Script, params: RunScriptParams): Promise<RunScriptResult> => {
+    const result = await script.transactionForDeployment(signer, params)
+    await signer.submitTransaction(result.unsignedTx, result.txId)
+    const confirmed = await waitTxConfirmed(signer.provider, result.txId)
+    return {...result, blockHash: confirmed.blockHash}
+  }
+
+  const setEnvironment = (key: string, value: string) => {
+    environments.set(key, value)
+  }
+
+  const getEnvironment = (key: string) => {
+    const value = environments.get(key)
+    if (typeof value === 'undefined') {
+      throw new Error(`${key} does not exist`)
+    }
+    return value
+  }
+
+  return {
+    provider: signer.provider,
+    account: account,
+    deployContract: deployContract,
+    runScript: runScript,
+    setEnvironment: setEnvironment,
+    getEnvironment: getEnvironment
```

Comment:
For deployment, I am wondering if it is a good pattern to have multiple scripts running separately and use `setEnvironment` & `getEnvironment` to pass information across the scripts, or is it better to just have a "deploy" script that deploys all contracts :thinking: 

Reference: https://stackoverflow.com/questions/69812460/how-to-deploy-multiple-smart-contracts-using-hardhat-deploy

Reply:
Both hardhat and truffle support multiple scripts running separately, I guess the benefit is to continue execution from the last failed step, the truffle will deploy a `Migration` contract which is used to record successful steps, and for the hardhat deploy plugin, this information will be saved in `.migration.json`.

But yes, the `setEnvironment` & `getEnvironment` is not friendly for developers, truffle/hardhat will save the contract deploy result once the `deployer.deploy()` execute successfully, so the developer doesn't need to call the `setEnvironment`, I'm working on improvements to support this.

Reply:
I pushed a few commits, now the DevX is similar to hardhat and truffle, users can get the contract deployment results by ContractId:

1. hardhat examples: https://github.com/wighawag/hardhat-deploy#deploy-scripts-tags-and-dependencies
2. truffle examples: https://github.com/wormhole-foundation/wormhole/blob/dev.v2/ethereum/migrations/3_deploy_bridge.js

But if the contract is created by another contract:

1. for `createSubContract`, we can get the contract id by `subContractId`
2. for `createContract`, we need to get the transaction output, I think we can add some utility functions to SDK to make it easier